website/docs/r/spot_fleet_request.html.markdown

```diff
@@ -78,6 +78,91 @@ resource "aws_spot_fleet_request" "foo" {
 }
 ```
 
+### Example with a dynamic block
+
+```hcl
+resource "aws_spot_fleet_request" "cheap_compute" {
+  iam_fleet_role                      = "${aws_iam_role.aws-ec2-spot-fleet-tagging-role.arn}"
+  target_capacity                     = var.target_capacity
+  valid_until                         = "${timeadd(timestamp(), var.timeadd)}"
+  allocation_strategy                 = var.allocation_strategy
+  fleet_type                          = var.fleet_type
+  wait_for_fulfillment                = true
+  terminate_instances_with_expiration = true
+
+  launch_specification {
+    instance_type            = "m4.10xlarge"
+    ami                      = "ami-1234"
+    spot_price               = "2.793"
+    placement_tenancy        = "dedicated"
+    iam_instance_profile_arn = "${aws_iam_instance_profile.example.arn}"
+  }
+
+  launch_specification {
+    for_each = [for i in local.instances_x_subnets : {
+      instance_type = i[0]
+      subnet_id     = i[1]
+    }]
+    content {
+      ami                         = var.ami_id
+      instance_type               = launch_specification.value.instance_type
+      subnet_id                   = launch_specification.value.subnet_id
+      vpc_security_group_ids      = var.vpc_security_group_ids
+      iam_instance_profile        = var.instance_profile_name
+      associate_public_ip_address = var.associate_public_ip_address
+      user_data                   = var.user_data
+      ebs_optimized               = var.ebs_optimized
+
+      root_block_device {
+        volume_size           = "8"
+        volume_type           = "gp2"
+        delete_on_termination = "true"
+      }
+
+      tags = {
+        Name        = var.name
+        tag_builder = "builder"
+      }
+    }
+  }
+}
+
+# merge two lists into a larger map
+locals {
+  instances_x_subnets = "${setproduct(var.instance_type, var.subnets)}"
```

Comment:
To make this example actually usable in isolation, it would be nice to include an actual `aws_vpc` and pair of `aws_subnet` resources here, and then perhaps move this expression up into the actual `for_each` argument above. We tend to try to keep these examples relatively simple and self-contained so that they don't require expert understanding of many different Terraform features, so I'd lean towards not using a local value here just to avoid introducing that more advanced feature here without any context.

Reply:
the reason I put this as resource is for the case a user wants to do an output of the merge (I am doing this).
Having it in line makes it opach and impossible to debug

Reply:
I agree that in a "real" configuration it would be better to put something this complicated out in a local value. Our resource type examples have a different goal, though: they are primarily aimed at showing the use of one resource type, so we prefer to keep anything else needed by the example as minimal as possible. In other words, they are not intended as something to be used in production exactly as written, but rather as a starting point to understand just this one resource type.

Now that I look at this again, I wonder if also we should try to simplify this to just be a per-subnet example, since introducing the `setproduct` pattern here feels a little heavy too. If it were just `var.subnets` we were iterating over then it would still show the general idea of dynamically generating the blocks without also introducing the concepts of local values and a function that the reader may not be familiar with yet.

Reply:
Do we ever provide advanced examples?
I'm not opposed to dropping setproduct, but it will make everyone else have to look up this solution (and afaik this is the best way to do this).
And as stated before, every single person using this resource will need this method anyway. 

Maybe have a note to link to setproduct page and just use a one level cycle? 
Trying to find a decent compromise that saves others hours of looking up how to implement for_each

Reply:
That is a good question and I think can reduce to this question: How does a new Terraform user learn that `setproduct` is a good way to iterate over multiple things at once?

My initial instinct for that is to mention it in [the main `dynamic` block documentation](https://www.terraform.io/docs/configuration/expressions.html#dynamic-blocks). That documentation already describes how to iterate over a single collection, so it seems like a logical extension to have a follow-up paragraph something like this:

```markdown
If you need to iterate over combinations of values from multiple collections,
use [`setproduct`](./functions/setproduct.html) to create a single collection
containing all of the combinations.
```

That way the reader can learn about it as part of learning about `dynamic` blocks in general, rather than only if they happen to also be working with AWS Spot Fleet requests.

To tie this all together, I'd then add an extra paragraph under the example to explain the meaning of that `dynamic` block, since there's no way to avoid introducing _that_ in this example:

```markdown
In this example, we use a
[`dynamic` block](/docs/configuration/expressions.html#dynamic-blocks)
to define zero or more `launch_specification` blocks, producing one for each
element in the list of subnet ids.
```

I think that would meet the goal of allowing the reader to find where to learn more about this new `dynamic` block feature (that they may have otherwise not seen before) and then we can gather information about different patterns for using it in the centralized documentation so that it's easier to discover along other learning paths too.

(The `dynamic` block documentation is over in the Terraform Core repository, so it's not possible to update that here in the same PR, but I'm happy to make that edit over in the core repository to complete this if that seems reasonable to you.)